server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -380,6 +352,32 @@ private void uploadNewSegments(Collection<String> localSegmentsPostRefresh, Acti
         GroupedActionListener<Void> batchUploadListener = new GroupedActionListener<>(mappedListener, filteredFiles.size());
 
         for (String src : filteredFiles) {
+            // Initializing listener here to ensure that the stats increment operations are thread-safe
+            UploadListener statsListener = new UploadListener() {
+                private long uploadStartTime = 0;
+
+                @Override
+                public void beforeUpload(String file) {
+                    // Start tracking the upload bytes started
+                    segmentTracker.addUploadBytesStarted(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
+                    uploadStartTime = System.currentTimeMillis();
+                }
+
+                @Override
+                public void onSuccess(String file) {
+                    // Track upload success
+                    segmentTracker.addUploadBytesSucceeded(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
+                    segmentTracker.addToLatestUploadedFiles(file);
+                    segmentTracker.addTotalUploadTimeInMs(Math.max(1, System.currentTimeMillis() - uploadStartTime));
+                }
+
+                @Override
+                public void onFailure(String file) {
+                    // Track upload failure
+                    segmentTracker.addUploadBytesFailed(segmentTracker.getLatestLocalFileNameLengthMap().get(file));
+                    segmentTracker.addTotalUploadTimeInMs(Math.max(1, System.currentTimeMillis() - uploadStartTime));
+                }
```

Comment:
lets have a method for creating the upload listener.

Reply:
Done. Addressed in the latest commit